### ISSUE TITLE
Vector Indexing Improvements

### ISSUE THREAD

**Issue by friedmud** [libMesh/libmesh#2920]
After doing some brainstorming with @loganharbour and @roystgnr ... I think that we could enable vectors to be a lot more efficient - in both time and space.

Here are the steps:
1. Do our own ghosting internal to Vectors (using `pushParallelVectorData`)
2. Share the global to map (probably put it in `DofMap`)
3. Remove `current_local_solution`
4. Add local indexing APIs to the `DofMap` / `DofObjects` (powered by the global to local map)
5. Store local dof indeces on `DofObjects` - allowing removal of the global to local map
6. Investigate optional deleting of ghost values during linear solves

Benefits:
- Doing our own ghosting:
  - Removes redundant memory usage in PETSc
  - Allows for the removal of `current_local_solution`
  - Allows for more fine-grained control over when ghosts are allocated (and deallocated!) and when they're updated
  - Add ghosting to `DistributedVector` for all types of solvers
- Working local indices
  - Let's us avoid expensive map hits
  - Allows removal of the global to local map
  - Provides a more streamlined interface for direct solution vector (for vector variables and array variables)

**Comment by roystgnr** [comment]
I think I'd feel safer about all this if we were to define a "standard libMesh benchmarking suite" as a prerequisite step 0.  Usually when I'm doing optimization work I'm making changes that I expect to be pure optimizations, so I pick whatever application I expect to hit the troublesome code path most often and I go to town, but here I worry that some of these steps are going to be tradeoffs, and if that turns out to be the case then it would be invaluable to be able to see what the best case/worst case/"weighted average" case results look like, for both CPU and memory use, so that we can make sure the tradeoffs are a net win.

> 2. Share the global to local map (probably put it in `DofMap`)

I think we can get a lot of mileage out of replacing our `unordered_map` usage with a more efficient hash map implementation, and we'll want to do that eventually because of all the other `unordered_map` uses in libMesh (including DistributedMesh internal storage!); we might revisit whether 2. is worthwhile after that's done.  It'll still be an improvement but it might be less of one.

> 3. Remove current_local_solution

Well, the hope is that for backwards compatibility it could be just another pointer to System::solution, in the near-term.

I've thought that current_local_solution was a bit of an embarrassment ever since we got ghosted-vector support added, but the backwards-compatibility issues are what always prevented me from taking a crack at eliminating it.

@friedmud says he tried to eliminate it at one point, but ran into the problem that "doing our own ghosting" should fix: when using a ghosted vector as the solution to a PETSc solve, PETSc would try to keep ghost values in sync for every matvec, would copy ghosted structure and metadata into krylov subspace vectors, and would generally be less efficient about the whole process.

> 4. Add local indexing APIs to the DofMap / DofObjects (powered by the global to local map)

I'm thinking more about NumericVector, at first.  For cases like MOOSE array variables where you've got this roundabout "global->local->global->NumericVector" path it'd be good to get rid of that third step (which would in turn get rid of the corresponding "->local" step inside NumericVector).

**Comment by YaqiWang** [comment]
This is almost exactly what I want. Thanks for proposing this. We do not have to do `removal of current_local_solution` first I guess. We can keep `GHOSTED` type even, maybe just add a new type like `PARALLELGOSTED`. In my own code long ago, I have mesh and can build a so-called `L2Variable` or `C1Variable` from the mesh which contains the info for communication, then use standard array for storing the solutions, which can be interpreted with a variable type. It is a little different from DofMap but similar idea.

**Comment by friedmud** [comment]
Still brainstorming...

We want to end up with _one_ contiguous array that holds both the locals _and_ the ghosts.  This is so that we can quickly and easily index into it using local indices.

This is what we already get right now with the PETSc local form for ghosted vectors.

To enable that - we're going to need to be a bit tricky.  We will need to create our own array (probably in `DistributedVector`) that is `num_local + num_ghosted` long.  Then, with PETSc, we'll need to use [VecCreateMPIWithArray](https://www.mcs.anl.gov/petsc/petsc-current/docs/manualpages/Vec/VecCreateMPIWithArray.html) to tell PETSc to use that vector.

PETSc will only be operating on the local part of the array (and won't know that it's secretly longer).  We'll keep the ghosted part up to date.

One good part about this is that all of the code we've written in MOOSE that utilizes PETSc's local form... will continue to work.

**Comment by fdkong** [comment]
>To enable that - we're going to need to be a bit tricky. We will need to create our own array (probably in DistributedVector) that is num_local + num_ghosted long. Then, with PETSc, we'll need to use VecCreateMPIWithArray to tell PETSc to use that vector.

Yes, it is very right. When PETSc duplicates Krylov subspaces, it will use only local parts because it can not see the "ghosted" elements.  

We should be able to remove global_to_local map from libmesh totally if we are going to use local indices ....
